{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Differential Expression and LogFC Calculation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import lines\n",
    "import anndata as ad\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "import scanpy as sc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "NUM_CPUS = 8\n",
    "LOG2_FC_THRESH = np.log2(2.0)\n",
    "NLOG10_PADJ_THRESH = -1*np.log10(0.05)\n",
    "\n",
    "# DATA_PATH = '/data/expression_atlas/v1/GSE122459'\n",
    "# DATA_PATH = '/data/expression_atlas/v1/GSE110914/'\n",
    "DATA_PATH = '/data/expression_atlas/v1/GSE162828/'\n",
    "# DATA_PATH = '/data/expression_atlas/v1/GSE102371/'\n",
    "\n",
    "RESULTS_PATH = '' + 'results/%s' % DATA_PATH.rstrip('/').split('/')[-1]\n",
    "\n",
    "DDS_TRANSCRIPT_FH = '' + 'results/%s_dds_transcript.h5_ad' % DATA_PATH.rstrip('/').split('/')[-1]\n",
    "DDS_GENE_FH = '' + 'results/%s_dds_gene.h5_ad' % DATA_PATH.rstrip('/').split('/')[-1]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read in previously created dds objects, should be filtered and have correct\n",
    "# metadata defined in obs.\n",
    "\n",
    "ad_dds = ad.read_h5ad(DDS_TRANSCRIPT_FH)\n",
    "ad_dds_gene = ad.read_h5ad(DDS_GENE_FH)\n",
    "\n",
    "dds = DeseqDataSet( \n",
    "                adata=ad_dds, \n",
    "                design_factors=[ \n",
    "                            c for c in ad_dds.obs.columns if c.startswith('condition')\n",
    "                    ]\n",
    "            )\n",
    "dds_gene = DeseqDataSet( \n",
    "                adata=ad_dds_gene, \n",
    "                design_factors=[ \n",
    "                            c for c in ad_dds_gene.obs.columns if c.startswith('condition')\n",
    "                    ]\n",
    "            )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.X, dds_gene.X"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.obs,dds_gene.obs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Manually define contrasts given conditions in metadata dataframe.\n",
    "\n",
    "for c in [c for c in dds.obs.columns if c.startswith('condition')]:\n",
    "    print(dds.obs[c].unique())\n",
    "\n",
    "# Pydeseq2 contrasts require condition-name, treatment level, reference level format.\n",
    "contrasts = {\n",
    "    'SLE_v_control': ['condition-1','TREAT-1','CONTROL'],\n",
    "    }\n",
    "\n",
    "# contrasts = {\n",
    "#     'T1D_v_control': ['condition-1','TREAT-1','CONTROL'],\n",
    "#     'preT1D_v_control': ['condition-1','TREAT-2','CONTROL']\n",
    "#     }\n",
    "\n",
    "# contrasts = {\n",
    "#     'SLE_v_control': ['condition-1','TREAT-1','CONTROL']\n",
    "#     }\n",
    "\n",
    "# contrasts = {\n",
    "#     'T1D_v_control': ['condition-1','TREAT-1','CONTROL'],\n",
    "#     }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dds.obsm['design_matrix'], dds_gene.obsm['design_matrix']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create Stats object. Define relevant contrasts for DE and LogFC computations and run tests. \n",
    "\n",
    "# Holds all DeseqStats objects as defined in contrasts.\n",
    "stat_results = {}\n",
    "\n",
    "for k, v in contrasts.items():\n",
    "\n",
    "    stat_res = DeseqStats(dds, contrast=v, n_cpus=NUM_CPUS)\n",
    "    stat_res_gene = DeseqStats(dds_gene, contrast=v, n_cpus=NUM_CPUS)\n",
    "\n",
    "    stat_res.summary()\n",
    "    stat_res_gene.summary()\n",
    "\n",
    "    stat_res.lfc_shrink()\n",
    "    stat_res_gene.lfc_shrink()\n",
    "    \n",
    "    stat_results['%s_transcript' % k] = stat_res\n",
    "    stat_results['%s_gene' % k] = stat_res_gene\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create volcano plots of DE-transcripts and -genes.\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),2,figsize=(10,5*len(contrasts)))\n",
    "\n",
    "ax = ax.reshape((-1,2,))\n",
    "\n",
    "scale_marker=2\n",
    "\n",
    "for i, k in enumerate(contrasts.keys()):\n",
    "    \n",
    "    kg = '%s_gene' % k\n",
    "    kt = '%s_transcript' % k\n",
    "\n",
    "    stat_results[kt].results_df['-log10_padj'] = -1. * np.log10(stat_results[kt].results_df['padj'])\n",
    "    stat_results[kg].results_df['-log10_padj'] = -1. * np.log10(stat_results[kg].results_df['padj'])\n",
    "\n",
    "    ax[i,0].scatter(\n",
    "            stat_results[kt].results_df['log2FoldChange'], \n",
    "            stat_results[kt].results_df['-log10_padj'], \n",
    "            alpha=0.1,\n",
    "            s=scale_marker*np.log2(stat_results[kt].results_df['baseMean']),\n",
    "            c=[ \n",
    "                '#1f77b4' if (abs(lf) > LOG2_FC_THRESH and nlp > NLOG10_PADJ_THRESH) else \n",
    "                '#ff7f0e' for i, (lf, nlp) in stat_results[kt].results_df[['log2FoldChange','-log10_padj']].iterrows()\n",
    "            ],\n",
    "        )\n",
    "\n",
    "    ax[i,1].scatter(\n",
    "            stat_results[kg].results_df['log2FoldChange'], \n",
    "            stat_results[kg].results_df['-log10_padj'], \n",
    "            alpha=0.05,\n",
    "            s=scale_marker*np.log2(stat_results[kg].results_df['baseMean']),\n",
    "            c=[ \n",
    "                '#1f77b4' if (abs(lf) > LOG2_FC_THRESH and nlp > NLOG10_PADJ_THRESH) else \n",
    "                '#ff7f0e' for i, (lf, nlp) in stat_results[kg].results_df[['log2FoldChange','-log10_padj']].iterrows()\n",
    "            ],\n",
    "        )\n",
    "\n",
    "    ax[i,0].set_xlabel('log2 FC')\n",
    "    ax[i,1].set_xlabel('log2 FC')\n",
    "    ax[i,0].set_ylabel('-log10 padj')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript' % k)\n",
    "    ax[i,1].set_title('%s Gene' % k)\n",
    "    \n",
    "    element_range = np.rint(np.linspace(\n",
    "                        1,\n",
    "                        5*round(max(np.log2(stat_results[kg].results_df['baseMean']))/5),\n",
    "                        4, \n",
    "                    ))\n",
    "\n",
    "    legend_elements = [lines.Line2D(\n",
    "                            [0], \n",
    "                            [0], \n",
    "                            lw=0, \n",
    "                            marker=\"o\", \n",
    "                            linestyle=None, \n",
    "                            markersize=(scale_marker*s)**0.5,\n",
    "                        ) for s in element_range]\n",
    "\n",
    "    legend = ax[i,1].legend(\n",
    "                    legend_elements,\n",
    "                    element_range,\n",
    "                    frameon=False, \n",
    "                    loc='upper left', \n",
    "                    bbox_to_anchor=(1.,1.),\n",
    "                    title='log2 mean expression'\n",
    "                )\n",
    "    ax[i,1].add_artist(legend)\n",
    "    \n",
    "    color_legend = ax[i,1].legend(\n",
    "                [\n",
    "                    lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#1f77b4'),\n",
    "                    lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#ff7f0e'),\n",
    "                    ],\n",
    "                [\n",
    "                    f'log2FC > {LOG2_FC_THRESH} and -log10_padj > {NLOG10_PADJ_THRESH:.2f}',\n",
    "                    f'log2FC < {LOG2_FC_THRESH} and -log10_padj < {NLOG10_PADJ_THRESH:.2f}',\n",
    "                    ],\n",
    "                frameon=False,\n",
    "                loc='upper left',\n",
    "                bbox_to_anchor=(1.,0.5,),\n",
    "                )   \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create MA plots of DE-transcripts and -genes.\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),2,figsize=(10,5*len(contrasts)))\n",
    "\n",
    "ax = ax.reshape((-1,2,))\n",
    "\n",
    "for i,k in enumerate(contrasts.keys()):\n",
    "\n",
    "    kg = '%s_gene' % k\n",
    "    kt = '%s_transcript' % k\n",
    "    \n",
    "    ax[i,0].scatter(\n",
    "            np.log2(stat_results[kt].results_df['baseMean']), \n",
    "            stat_results[kt].results_df['log2FoldChange'], \n",
    "            alpha=0.1,\n",
    "            s=stat_results[kt].results_df['-log10_padj'],\n",
    "            c=[ \n",
    "                '#1f77b4' if (abs(lf) > LOG2_FC_THRESH and nlp > NLOG10_PADJ_THRESH) else \n",
    "                '#ff7f0e' for i, (lf, nlp) in stat_results[kt].results_df[['log2FoldChange','-log10_padj']].iterrows()\n",
    "            ],\n",
    "        )\n",
    "\n",
    "    ax[i,1].scatter(\n",
    "            np.log2(stat_results[kg].results_df['baseMean']), \n",
    "            stat_results[kg].results_df['log2FoldChange'], \n",
    "            alpha=0.1,\n",
    "            s=stat_results[kg].results_df['-log10_padj'],\n",
    "            c=[ \n",
    "                '#1f77b4' if (abs(lf) > LOG2_FC_THRESH and nlp > NLOG10_PADJ_THRESH) else \n",
    "                '#ff7f0e' for i, (lf, nlp) in stat_results[kg].results_df[['log2FoldChange','-log10_padj']].iterrows()\n",
    "            ],\n",
    "        )\n",
    "\n",
    "    ax[i,0].set_xlabel('log2 mean expression')\n",
    "    ax[i,1].set_xlabel('log2 mean expression')\n",
    "    ax[i,0].set_ylabel('log2 FC expression')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript' % k)\n",
    "    ax[i,1].set_title('%s Gene' % k)\n",
    "\n",
    "\n",
    "    element_range = np.rint(np.linspace(\n",
    "                    1,\n",
    "                    5*round(max(stat_results[kg].results_df['-log10_padj'])/5),\n",
    "                    4, \n",
    "                ))\n",
    "\n",
    "    legend_elements = [lines.Line2D(\n",
    "                            [0], \n",
    "                            [0], \n",
    "                            lw=0, \n",
    "                            marker=\"o\", \n",
    "                            linestyle=None, \n",
    "                            markersize=s**0.5,\n",
    "                        ) for s in element_range]\n",
    "\n",
    "    legend = ax[i,1].legend(\n",
    "                    legend_elements,\n",
    "                    element_range,\n",
    "                    frameon=False, \n",
    "                    loc='upper left', \n",
    "                    bbox_to_anchor=(1.,1.),\n",
    "                    title='-log10_padj'\n",
    "                )\n",
    "    ax[i,1].add_artist(legend)\n",
    "    \n",
    "    color_legend = ax[i,1].legend(\n",
    "                [\n",
    "                    lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#1f77b4'),\n",
    "                    lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#ff7f0e'),\n",
    "                    ],\n",
    "                [\n",
    "                    f'log2FC > {LOG2_FC_THRESH} and -log10_padj > {NLOG10_PADJ_THRESH:.2f}',\n",
    "                    f'log2FC < {LOG2_FC_THRESH} and -log10_padj < {NLOG10_PADJ_THRESH:.2f}',\n",
    "                    ],\n",
    "                frameon=False,\n",
    "                loc='upper left',\n",
    "                bbox_to_anchor=(1.,0.5,),\n",
    "                )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for i,k in enumerate(contrasts.keys()):\n",
    "\n",
    "    kg = '%s_gene' % k\n",
    "    kt = '%s_transcript' % k\n",
    "\n",
    "    markers = stat_results[kt].results_df[(abs(stat_results[kt].results_df['log2FoldChange']) > LOG2_FC_THRESH) & (stat_results[kt].results_df['-log10_padj'] > NLOG10_PADJ_THRESH)]\n",
    "    print('%s: %s'  % (kt, len(markers)))\n",
    "    print(markers.sort_values('log2FoldChange', axis=0).to_string())\n",
    "\n",
    "    markers = stat_results[kg].results_df[(abs(stat_results[kg].results_df['log2FoldChange']) > LOG2_FC_THRESH) & (stat_results[kg].results_df['-log10_padj'] > NLOG10_PADJ_THRESH)]\n",
    "    print('%s: %s' % (kg, len(markers)))\n",
    "    print(markers.sort_values('log2FoldChange', axis=0).to_string())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transfer gene to transcript mappings to dds.var dataframe. Plot LogFC between transcript- \n",
    "# and gene-level quantifications.\n",
    "\n",
    "gene_transcript_mapping = dict(zip(dds.uns['gene_transcript_mapping']['tx'],dds.uns['gene_transcript_mapping']['gene_id']))\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),1,figsize=(5,5*len(contrasts)))\n",
    "\n",
    "if type(ax) != np.ndarray:\n",
    "    ax = np.array(ax)\n",
    "\n",
    "ax = ax.reshape((-1,1,))\n",
    "\n",
    "for i,k in enumerate(contrasts.keys()):\n",
    "\n",
    "    kg = '%s_gene' % k\n",
    "    kt = '%s_transcript' % k\n",
    "    \n",
    "    stat_results[kt].results_df['gene_id'] = stat_results[kt].results_df.index.map(lambda x: gene_transcript_mapping[x])\n",
    "\n",
    "    df = stat_results[kt].results_df.merge(stat_results[kg].results_df, left_on='gene_id', right_on='gene_id')\n",
    "    ax[i,0].scatter(\n",
    "            df['log2FoldChange_x'], \n",
    "            df['log2FoldChange_y'],\n",
    "            alpha=0.05,\n",
    "            s=5*df['-log10_padj_x']\n",
    "        )\n",
    "    \n",
    "    ax[i,0].set_xlabel('log2 FC transcript')\n",
    "    ax[i,0].set_ylabel('log2 FC gene')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript v Gene logFC' % k)\n",
    "\n",
    "    element_range = np.rint(np.linspace(\n",
    "                        1,\n",
    "                        5*round(max(stat_results[kg].results_df['-log10_padj'])/5),\n",
    "                        4, \n",
    "                    ))\n",
    "\n",
    "    legend_elements = [lines.Line2D(\n",
    "                            [0], \n",
    "                            [0], \n",
    "                            lw=0, \n",
    "                            marker=\"o\", \n",
    "                            linestyle=None, \n",
    "                            markersize=s**0.5,\n",
    "                        ) for s in element_range]\n",
    "\n",
    "    legend = ax[i,0].legend(\n",
    "                    legend_elements,\n",
    "                    element_range,\n",
    "                    frameon=False, \n",
    "                    loc='upper left', \n",
    "                    bbox_to_anchor=(1.,1.),\n",
    "                    title='-log10_padj transcript'\n",
    "                )\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dump results dataframes to results folder.\n",
    "\n",
    "for k in contrasts.keys():\n",
    "\n",
    "    kt = '%s_transcript' % k\n",
    "    kg = '%s_gene' % k\n",
    "\n",
    "    stat_results[kt].results_df.to_csv('%s_%s_%s.csv' % (RESULTS_PATH, 'transcript', k))\n",
    "    stat_results[kg].results_df.to_csv('%s_%s_%s.csv' % (RESULTS_PATH, 'gene', k))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
