{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 3. Differential Expression Visualization and QC\n",
    "\n",
    "This notebook performs visualization and quality control analysis of differential expression results.\n",
    "\n",
    "#### Workflow Overview\n",
    "1. Creation of standard DE visualization plots:\n",
    "   - Volcano plots\n",
    "   - MA plots\n",
    "   - Gene vs transcript level comparisons\n",
    "\n",
    "2. Quality control of pydeseq2 results:\n",
    "   - Validation of log fold change calculations\n",
    "   - Comparison of raw vs squeezed log fold changes\n",
    "   - Cross-contrast comparisons\n",
    "\n",
    "3. Data Processing:\n",
    "   - TPM scaling with normalized counts\n",
    "   - Calculating group averages \n",
    "\n",
    "#### Expected Input Files\n",
    "-The notebook expects DESeq2 dataset objects (h5ad format) from the previous notebook containing:\n",
    "   - Normalized counts\n",
    "   - Size factors\n",
    "   - Differential expression results\n",
    "   - Sample metadata\n",
    "   - Contrast definitions\n",
    "\n",
    "#### Output Files\n",
    "- CSV files containing DE results for each contrast\n",
    "- CSV file that describes samples used in each contrast\n",
    "- Updated AnnData objects"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import warnings\n",
    "from itertools import combinations\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import anndata as ad\n",
    "import scanpy as sc\n",
    "from pybiomart import Dataset\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import lines\n",
    "\n",
    "from pydeseq2.dds import DeseqDataSet\n",
    "from pydeseq2.ds import DeseqStats\n",
    "\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.1 Configure Notebook\n",
    "\n",
    "#### Required Variable Definitions:\n",
    "Processing Parameters\n",
    "- *NUM_CPUS*: Number of CPUs to use for parallel processing (default: 8)\n",
    "- *LOG2_FC_THRESH*: Log2 fold change threshold for significance (default: log2(2.0))\n",
    "- *NLOG10_PADJ_THRESH*: -log10 adjusted p-value threshold (default: -log10(0.05))\n",
    "\n",
    "File Paths\n",
    "- *DATA_PATH*: Root directory path for the project (default: current working directory)\n",
    "- *RESULTS_PATH*: Output directory for differential expression results (default: DATA_PATH/de_results)\n",
    "- *DDS_TRANSCRIPT_FH*: Path to transcript-level DESeq2 dataset object\n",
    "- *DDS_GENE_FH*: Path to gene-level DESeq2 dataset object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "NUM_CPUS = 8\n",
    "LOG2_FC_THRESH = np.log2(2.0)\n",
    "NLOG10_PADJ_THRESH = -1*np.log10(0.05)\n",
    "\n",
    "DATA_PATH = Path.cwd().parent\n",
    "\n",
    "EXPERIMENT_ID = DATA_PATH.parts[-1]\n",
    "\n",
    "RESULTS_PATH = DATA_PATH / 'de_results'\n",
    "\n",
    "DDS_TRANSCRIPT_FH = RESULTS_PATH / f'{EXPERIMENT_ID}_dds_transcript.h5_ad'\n",
    "DDS_GENE_FH = RESULTS_PATH / f'{EXPERIMENT_ID}_dds_gene.h5_ad'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Read in previously created dds objects, should be filtered and have correct\n",
    "# metadata defined in obs.\n",
    "\n",
    "dds = ad.read_h5ad(DDS_TRANSCRIPT_FH)\n",
    "dds_gene = ad.read_h5ad(DDS_GENE_FH)\n",
    "\n",
    "contrasts = dds.uns['contrasts']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.2 Fetch ensembl gene id - external gene mappings from biomart."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Fetch ensembl gene id - external gene name mappings from biomart.\n",
    "\n",
    "species_mapping = 'hsapiens' if any(dds.var.index.str.startswith('ENSG')) else (\n",
    "    'mmusculus' if any(dds.var.index.str.startswith('ENSMUSG')) else None\n",
    ")\n",
    "if not species_mapping:\n",
    "    raise Exception('Must pick a valid species, adjust above if not human/mouse.')\n",
    "\n",
    "dataset = Dataset(\n",
    "    name=species_mapping,\n",
    "    host='http://www.ensembl.org',\n",
    ")\n",
    "\n",
    "external_gene_mapping = dataset.query(\n",
    "    attributes=['ensembl_gene_id', 'external_gene_name']\n",
    ")\n",
    "\n",
    "external_gene_mapping.rename({'Gene stable ID': 'gene_id', 'Gene name': 'gene_name'}, axis=1, inplace=True)\n",
    "\n",
    "external_gene_mapping"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.3 Volcano plots for all contrasts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create volcano plots of DE-transcripts and -genes.\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts), 2, figsize=(10,5*len(contrasts)))\n",
    "\n",
    "ax = ax.reshape((-1,2))\n",
    "\n",
    "scale_marker=2\n",
    "\n",
    "for i, k in enumerate(contrasts.keys()):\n",
    "    \n",
    "    ax[i,0].scatter(\n",
    "        dds.uns['stat_results'][k]['log2FoldChange'], \n",
    "        dds.uns['stat_results'][k]['-log10_padj'], \n",
    "        alpha=0.1,\n",
    "        s=scale_marker*np.log2(dds.uns['stat_results'][k]['baseMean']),\n",
    "        c=dds_gene.uns['stat_results'][k].apply(\n",
    "            lambda x: \n",
    "                '#1f77b4' if (abs(x['log2FoldChange']) > LOG2_FC_THRESH) and (x['-log10_padj'] > NLOG10_PADJ_THRESH) else '#ff7f0e'\n",
    "            axis=1,\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    ax[i,1].scatter(\n",
    "        dds_gene.uns['stat_results'][k]['log2FoldChange'], \n",
    "        dds_gene.uns['stat_results'][k]['-log10_padj'], \n",
    "        alpha=0.05,\n",
    "        s=scale_marker*np.log2(dds_gene.uns['stat_results'][k]['baseMean']),\n",
    "        c=dds_gene.uns['stat_results'][k].apply(\n",
    "            lambda x: \n",
    "                '#1f77b4' if (abs(x['log2FoldChange']) > LOG2_FC_THRESH) and (x['-log10_padj'] > NLOG10_PADJ_THRESH) else '#ff7f0e'\n",
    "            axis=1,\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    ax[i,0].set_xlabel('log2 FC')\n",
    "    ax[i,1].set_xlabel('log2 FC')\n",
    "    ax[i,0].set_ylabel('-log10 padj')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript' % k)\n",
    "    ax[i,1].set_title('%s Gene' % k)\n",
    "    \n",
    "    element_range = np.rint(\n",
    "        np.linspace(1, 5*round(max(np.log2(dds_gene.uns['stat_results'][k]['baseMean']))/5), 4)\n",
    "    )\n",
    "\n",
    "    legend_elements = [\n",
    "        lines.Line2D(\n",
    "            [0], \n",
    "            [0], \n",
    "            lw=0, \n",
    "            marker=\"o\", \n",
    "            linestyle=None, \n",
    "            markersize=(scale_marker*s)**0.5,\n",
    "        ) for s in element_range\n",
    "    ]\n",
    "\n",
    "    legend = ax[i,1].legend(\n",
    "        legend_elements,\n",
    "        element_range,\n",
    "        frameon=False, \n",
    "        loc='upper left', \n",
    "        bbox_to_anchor=(1.,1.),\n",
    "        title='log2 mean expression'\n",
    "    )\n",
    "    ax[i,1].add_artist(legend)\n",
    "    \n",
    "    color_legend = ax[i,1].legend(\n",
    "        [\n",
    "            lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#1f77b4'),\n",
    "            lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#ff7f0e'),\n",
    "        ],\n",
    "        [\n",
    "            f'log2FC > {LOG2_FC_THRESH} and -log10_padj > {NLOG10_PADJ_THRESH:.2f}',\n",
    "            f'log2FC < {LOG2_FC_THRESH} and -log10_padj < {NLOG10_PADJ_THRESH:.2f}',\n",
    "        ],\n",
    "        frameon=False,\n",
    "        loc='upper left',\n",
    "        bbox_to_anchor=(1.,0.5),\n",
    "    )   \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.4 MA plots for all contrasts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create MA plots of DE-transcripts and -genes.\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),2,figsize=(10,5*len(contrasts)))\n",
    "\n",
    "ax = ax.reshape((-1,2,))\n",
    "\n",
    "for i,k in enumerate(contrasts.keys()):\n",
    "\n",
    "    ax[i,0].scatter(\n",
    "        np.log2(dds.uns['stat_results'][k]['baseMean']), \n",
    "        dds.uns['stat_results'][k]['log2FoldChange'], \n",
    "        alpha=0.1,\n",
    "        s=dds.uns['stat_results'][k]['-log10_padj'],\n",
    "        c=dds_gene.uns['stat_results'][k].apply(\n",
    "            lambda x: \n",
    "                '#1f77b4' if (abs(x['log2FoldChange']) > LOG2_FC_THRESH) and (x['-log10_padj'] > NLOG10_PADJ_THRESH) else '#ff7f0e'\n",
    "            axis=1,\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    ax[i,1].scatter(\n",
    "        np.log2(dds_gene.uns['stat_results'][k]['baseMean']), \n",
    "        dds_gene.uns['stat_results'][k]['log2FoldChange'], \n",
    "        alpha=0.1,\n",
    "        s=dds_gene.uns['stat_results'][k]['-log10_padj'],\n",
    "        c=dds_gene.uns['stat_results'][k].apply(\n",
    "            lambda x: \n",
    "                '#1f77b4' if (abs(x['log2FoldChange']) > LOG2_FC_THRESH) and (x['-log10_padj'] > NLOG10_PADJ_THRESH) else '#ff7f0e'\n",
    "            axis=1,\n",
    "        ),\n",
    "    )\n",
    "\n",
    "    ax[i,0].set_xlabel('log2 mean expression')\n",
    "    ax[i,1].set_xlabel('log2 mean expression')\n",
    "    ax[i,0].set_ylabel('log2 FC expression')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript' % k)\n",
    "    ax[i,1].set_title('%s Gene' % k)\n",
    "\n",
    "\n",
    "    element_range = np.rint(\n",
    "        np.linspace(1, 5*round(max(dds_gene.uns['stat_results'][k]['-log10_padj'])/5), 4)\n",
    "    )\n",
    "\n",
    "    legend_elements = [\n",
    "        lines.Line2D(\n",
    "            [0], \n",
    "            [0], \n",
    "            lw=0, \n",
    "            marker=\"o\", \n",
    "            linestyle=None, \n",
    "            markersize=s**0.5,\n",
    "        ) for s in element_range\n",
    "    ]\n",
    "\n",
    "    legend = ax[i,1].legend(\n",
    "        legend_elements,\n",
    "        element_range,\n",
    "        frameon=False, \n",
    "        loc='upper left', \n",
    "        bbox_to_anchor=(1.,1.),\n",
    "        title='-log10_padj'\n",
    "    )\n",
    "    ax[i,1].add_artist(legend)\n",
    "    \n",
    "    color_legend = ax[i,1].legend(\n",
    "        [\n",
    "            lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#1f77b4'),\n",
    "            lines.Line2D([0], [0], lw=0, marker='o', linestyle=None, markerfacecolor='#ff7f0e'),\n",
    "        ],\n",
    "        [\n",
    "            f'log2FC > {LOG2_FC_THRESH} and -log10_padj > {NLOG10_PADJ_THRESH:.2f}',\n",
    "            f'log2FC < {LOG2_FC_THRESH} and -log10_padj < {NLOG10_PADJ_THRESH:.2f}',\n",
    "        ],\n",
    "        frameon=False,\n",
    "        loc='upper left',\n",
    "        bbox_to_anchor=(1.,0.5),\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.5 DE genes/transcripts for each contrasts filtered by log2fc and padj thresholds."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Filter summary tables based upon thresholds.\n",
    "\n",
    "for i,k in enumerate(contrasts.keys()):\n",
    "\n",
    "    markers = dds.uns['stat_results'][k].loc[\n",
    "        (abs(dds.uns['stat_results'][k]['log2FoldChange']) > LOG2_FC_THRESH) & \n",
    "        (dds.uns['stat_results'][k]['-log10_padj'] > NLOG10_PADJ_THRESH)\n",
    "    ]\n",
    "    print('%s Transcript: %s'  % (k, len(markers)))\n",
    "    print(markers.sort_values('log2FoldChange', axis=0))\n",
    "\n",
    "    markers = dds_gene.uns['stat_results'][k].loc[\n",
    "        (abs(dds_gene.uns['stat_results'][k]['log2FoldChange']) > LOG2_FC_THRESH) & \n",
    "        (dds_gene.uns['stat_results'][k]['-log10_padj'] > NLOG10_PADJ_THRESH)\n",
    "    ]\n",
    "    print('%s Gene: %s' % (k, len(markers)))\n",
    "    print(markers.sort_values('log2FoldChange', axis=0))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.6 Plot raw log2fc vs. squeezed log2fc for each contrasts (qc issues with pydeseq2 log2fc calculations)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Run LFC calculations on raw inputs as gut-check for issues with pydeseq2 LFC calculations.\n",
    "\n",
    "for k, v in contrasts.items():\n",
    "\n",
    "    dds.varm['LFC_reflevel_%s_raw' % v[2]] = dds.varm['LFC_reflevel_%s' % v[2]].copy()\n",
    "    dds_gene.varm['LFC_reflevel_%s_raw' % v[2]] = dds_gene.varm['LFC_reflevel_%s' % v[2]].copy()\n",
    "\n",
    "    dds.varm['LFC_reflevel_%s_raw' % v[2]]['%s_%s_vs_%s' % tuple(v)] = np.log2(\n",
    "        np.mean(dds.X[dds.obs[v[0]] == v[1]],axis=0) / \n",
    "            np.mean(dds.X[dds.obs[v[0]] == v[2]], axis=0)\n",
    "    )\n",
    "    \n",
    "    dds_gene.varm['LFC_reflevel_%s_raw' % v[2]]['%s_%s_vs_%s' % tuple(v)] = np.log2(\n",
    "        np.mean(dds_gene.X[dds_gene.obs[v[0]] == v[1]],axis=0) / \n",
    "            np.mean(dds_gene.X[dds_gene.obs[v[0]] == v[2]], axis=0)\n",
    "    )\n",
    "                                                \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot raw LFC calculations vs. squeezed LFC calculations.\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),2,figsize=(10,5*len(contrasts)))\n",
    "\n",
    "ax = ax.reshape((-1,2,))\n",
    "\n",
    "for i,(k,v) in enumerate(contrasts.items()):\n",
    "    \n",
    "    ax[i,0].scatter(\n",
    "        dds.varm['LFC_reflevel_%s' % v[2]]['%s_%s_vs_%s' % tuple(v)],\n",
    "        dds.varm['LFC_reflevel_%s_raw' % v[2]]['%s_%s_vs_%s' % tuple(v)],\n",
    "        alpha=0.1,\n",
    "        s=0.1,\n",
    "    )\n",
    "    \n",
    "    ax[i,1].scatter(\n",
    "        dds_gene.varm['LFC_reflevel_%s' % v[2]]['%s_%s_vs_%s' % tuple(v)],\n",
    "        dds_gene.varm['LFC_reflevel_%s_raw' % v[2]]['%s_%s_vs_%s' % tuple(v)],\n",
    "        alpha=0.1,\n",
    "        s=0.1,\n",
    "    )\n",
    "\n",
    "    ax[i,0].set_xlabel('log2 FC squeezed')\n",
    "    ax[i,1].set_xlabel('log2 FC squeezed')\n",
    "    ax[i,0].set_ylabel('log2 FC raw')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript' % k)\n",
    "    ax[i,1].set_title('%s Gene' % k)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.7 Plot log2fc gene vs. log2fc transcript."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Transfer gene to transcript mappings to dds.var dataframe. Plot LogFC between transcript- \n",
    "# and gene-level quantifications.\n",
    "\n",
    "gene_transcript_mapping = dict(zip(dds.uns['gene_transcript_mapping']['tx'],dds.uns['gene_transcript_mapping']['gene_id']))\n",
    "\n",
    "fig, ax = plt.subplots(len(contrasts),1,figsize=(5,5*len(contrasts)))\n",
    "\n",
    "if type(ax) != np.ndarray:\n",
    "    ax = np.array(ax)\n",
    "\n",
    "ax = ax.reshape((-1,1,))\n",
    "\n",
    "for i,k in enumerate(contrasts.keys()):\n",
    "    \n",
    "    dds.uns['stat_results'][k]['gene_id'] = dds.uns['stat_results'][k].index.map(lambda x: gene_transcript_mapping[x])\n",
    "\n",
    "    df = dds.uns['stat_results'][k].merge(dds_gene.uns['stat_results'][k], left_on='gene_id', right_index=True)\n",
    "    ax[i,0].scatter(\n",
    "        df['log2FoldChange_x'], \n",
    "        df['log2FoldChange_y'],\n",
    "        alpha=0.05,\n",
    "        s=5*df['-log10_padj_x']\n",
    "    )\n",
    "    \n",
    "    ax[i,0].set_xlabel('log2 FC transcript')\n",
    "    ax[i,0].set_ylabel('log2 FC gene')\n",
    "\n",
    "    ax[i,0].set_title('%s Transcript v Gene logFC' % k)\n",
    "\n",
    "    element_range = np.rint(\n",
    "        np.linspace(1, 5*round(max(dds_gene.uns['stat_results'][k]['-log10_padj'])/5), 4)\n",
    "    )\n",
    "\n",
    "    legend_elements = [\n",
    "        lines.Line2D(\n",
    "            [0], \n",
    "            [0], \n",
    "            lw=0, \n",
    "            marker=\"o\", \n",
    "            linestyle=None, \n",
    "            markersize=s**0.5,\n",
    "        ) for s in element_range\n",
    "    ]\n",
    "\n",
    "    legend = ax[i,0].legend(\n",
    "        legend_elements,\n",
    "        element_range,\n",
    "        frameon=False, \n",
    "        loc='upper left', \n",
    "        bbox_to_anchor=(1.,1.),\n",
    "        title='-log10_padj transcript'\n",
    "    )\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.8 plot log2fc of all pairs of contrasts."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Plot LogFC between all combinations of differnt contrasts. \n",
    "\n",
    "contrast_combinations = [(k_1, k_2) for k_1, k_2 in combinations(contrasts.keys(), 2)]\n",
    "\n",
    "if len(contrast_combinations) > 0:\n",
    "\n",
    "    fig, ax = plt.subplots(len(contrast_combinations),2,figsize=(10,5*len(contrast_combinations)))\n",
    "\n",
    "    ax = ax.reshape((-1,2,))\n",
    "\n",
    "    for i, (k_1, k_2) in enumerate(contrast_combinations):\n",
    "\n",
    "        ax[i,0].scatter(\n",
    "            dds.uns['stat_results'][k_1]['log2FoldChange'], \n",
    "            dds.uns['stat_results'][k_2]['log2FoldChange'], \n",
    "            s=1, \n",
    "            alpha=0.1,\n",
    "        )\n",
    "        \n",
    "        ax[i,0].set_xlabel('log2 FC %s %s' % ('transcript',k_1))\n",
    "        ax[i,0].set_ylabel('log2 FC %s %s' % ('transcript',k_2))\n",
    "\n",
    "        \n",
    "        ax[i,1].scatter(\n",
    "            dds_gene.uns['stat_results'][k_1]['log2FoldChange'], \n",
    "            dds_gene.uns['stat_results'][k_2]['log2FoldChange'], \n",
    "            s=1, \n",
    "            alpha=0.1,\n",
    "        )\n",
    "        \n",
    "        ax[i,1].set_xlabel('log2 FC %s %s' % ('gene',k_1))\n",
    "        ax[i,1].set_ylabel('log2 FC %s %s' % ('gene',k_2))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.9 Create scaled normed_counts to attempt to shift normed_counts distribution to be similar in magnitude to tpm \n",
    "* For various reasons, this doesn't work well in practice"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Fit scaling factors for normed counts to scale to the tpm distribution. This is to make visualizing the \n",
    "# deseq2 normed counts easier across experiments.\n",
    "\n",
    "scale_factor_t = np.sum(\n",
    "    dds.layers['normed_counts'] * dds.layers['raw_tpm']) \n",
    "    / np.sum(dds.layers['normed_counts']**2\n",
    ")\n",
    "\n",
    "scale_factor_g = np.sum(\n",
    "    dds_gene.layers['normed_counts'] * dds_gene.layers['raw_tpm']) \n",
    "    / np.sum(dds_gene.layers['normed_counts']**2\n",
    ")    \n",
    "    \n",
    "dds.layers['normed_counts_transform'] = dds.layers['normed_counts']*scale_factor_t\n",
    "dds_gene.layers['normed_counts_transform'] = dds_gene.layers['normed_counts']*scale_factor_g\n",
    "\n",
    "dds.uns['normed_counts_transform_scalefactor'] = scale_factor_t\n",
    "dds_gene.uns['normed_counts_transform_scalefactor'] = scale_factor_g"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Wipe existing columns meannormedcounts in each output contrast if they exist.\n",
    "\n",
    "column_prefixes = dds.obs.columns[dds.obs.colums.str.startswith(('condition-', 'group-'))].tolist() \n",
    "    + dds_gene.obs.columns[dds_gene.obs.colums.str.startswith(('condition-', 'group-'))].tolist()\n",
    "                    [c for c in dds_gene.obs.columns if c.startswith('condition-') or c.startswith('group-')]\n",
    "\n",
    "for k in contrasts.keys():\n",
    "    dds.uns['stat_results'][k].drop(\n",
    "        [\n",
    "            c for c in dds.uns['stat_results'][k].columns if \\\n",
    "                any([True if c.startswith(cp) else False for cp in column_prefixes])\n",
    "        ],\n",
    "        axis=1,\n",
    "        inplace=True,\n",
    "    )\n",
    "    \n",
    "    dds_gene.uns['stat_results'][k].drop(\n",
    "        [\n",
    "            c for c in dds_gene.uns['stat_results'][k].columns if \\\n",
    "                any([True if c.startswith(cp) else False for cp in column_prefixes])\n",
    "        ],\n",
    "        axis=1,\n",
    "        inplace=True,\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3.10 Merge normed_count group summaries into dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Create group mean counts and individual sample counts that will be merged into de_dataframes.\n",
    "\n",
    "dds_gene.varm['normed_counts_group_mean'] = pd.DataFrame(index=dds_gene.var.index)\n",
    "dds.varm['normed_counts_group_mean'] = pd.DataFrame(index=dds.var.index)\n",
    "dds_gene.varm['normed_counts_group'] = pd.DataFrame(index=dds_gene.var.index)\n",
    "dds.varm['normed_counts_group'] = pd.DataFrame(index=dds.var.index)\n",
    "\n",
    "for c in dds_gene.obs.columns[dds_gene.obs.columns.str.startswith(('condition-', 'gene-'))]:\n",
    "    levels = dds_gene.obs[c].unique()\n",
    "    for l in levels:\n",
    "        l_i = np.where(np.isin(dds_gene.obs.index.values, dds_gene.obs[dds_gene.obs[c] == l].index.values,))\n",
    "        dds_gene.varm['normed_counts_group_mean']['%s_%s_meannormedcounts' % (c,l)] = \\\n",
    "            dds_gene.layers['normed_counts_transform'][l_i[0],:].mean(axis=0)\n",
    "        \n",
    "        dds_gene.varm['normed_counts_group'][['%s_%s_%s' % (c,l,s) for s in dds.obs.index[l_i]]] = \\\n",
    "            dds_gene.layers['normed_counts_transform'][l_i[0],:].T\n",
    "\n",
    "        l_i = np.where(np.isin(dds.obs.index.values, dds.obs[dds.obs[c] == l].index.values,))\n",
    "        dds.varm['normed_counts_group_mean']['%s_%s_meannormedcounts' % (c,l)] = \\\n",
    "            dds.layers['normed_counts_transform'][l_i[0],:].mean(axis=0)\n",
    "        \n",
    "        dds.varm['normed_counts_group'][['%s_%s_%s' % (c,l,s) for s in dds.obs.index[l_i]]] = \\\n",
    "            dds.layers['normed_counts_transform'][l_i[0],:].T\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Merge group-level means into stat_results dataframes.\n",
    "\n",
    "for k,v in contrasts.items():\n",
    "    \n",
    "    keep_cols_mean = [\n",
    "        c for c in dds.varm['normed_counts_group_mean'] \n",
    "        if c.startswith(v[0]) and c.split('_')[1] in (v[1],v[2],)\n",
    "    ]\n",
    "    \n",
    "    keep_cols_ind = [\n",
    "        c for c in dds.varm['normed_counts_group']\n",
    "        if c.startswith(v[0]) and c.split('_')[1] in (v[1],v[2],)\n",
    "    ]\n",
    "\n",
    "    dds.uns['stat_results'][k] = dds.uns['stat_results'][k].merge(\n",
    "        dds.varm['normed_counts_group_mean'][keep_cols_mean], \n",
    "        left_index=True, \n",
    "        right_index=True,\n",
    "    )\n",
    "    \n",
    "    dds.uns['stat_results'][k] = dds.uns['stat_results'][k].merge(\n",
    "        dds.varm['normed_counts_group'][keep_cols_ind],\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "    )\n",
    "        \n",
    "    dds_gene.uns['stat_results'][k] = dds_gene.uns['stat_results'][k].merge(\n",
    "        dds_gene.varm['normed_counts_group_mean'][keep_cols_mean], \n",
    "        left_index=True, \n",
    "        right_index=True,\n",
    "    )\n",
    "    \n",
    "    dds_gene.uns['stat_results'][k] = dds_gene.uns['stat_results'][k].merge(\n",
    "        dds_gene.varm['normed_counts_group'][keep_cols_ind],\n",
    "        left_index=True,\n",
    "        right_index=True,\n",
    "    )\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Define a contrast dataframe that outlines which condition/sample types were used in each contrast.\n",
    "\n",
    "contrast_df = pd.DataFrame(\n",
    "    columns=(\n",
    "        ['contrast','contrast_condition'] \n",
    "        + list(set([c[0] for c in dds.uns['contrasts'].values()]))\n",
    "    ),\n",
    ")\n",
    "\n",
    "for n, c in dds.uns['contrasts'].items():\n",
    "    contrast_df.loc[len(contrast_df),['contrast','contrast_condition',c[0]]] = n,c[0],c[1]\n",
    "    contrast_df.loc[len(contrast_df),['contrast','contrast_condition',c[0]]] = n,c[0],c[2]\n",
    "\n",
    "for c in contrast_df.columns[contrast_df.columns.str.startswith('condition-')]:\n",
    "    contrast_df = contrast_df.merge(dds.obs.reset_index(), on=c)\n",
    "\n",
    "contrast_df.to_csv(RESULTS_PATH / f'{EXPERIMENT_ID}_metadata_contrast.csv', index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# Dump results dataframes to results folder.\n",
    "\n",
    "for k in contrasts.keys():\n",
    "\n",
    "    dds.uns['stat_results'][k].to_csv(\n",
    "        RESULTS_PATH / f'{EXPERIMENT_ID}_transcript_{k}.csv'\n",
    "    )\n",
    "    dds_gene.uns['stat_results'][k].to_csv(\n",
    "        RESULTS_PATH / f'{EXPERIMENT_ID}_gene_{k}.csv'\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write dds objects to files for gsea analysis.\n",
    "\n",
    "# Pydeseq2 supports trend_coeffs/replaced as either np.array or pd.series, np.array required for \n",
    "# saving h5-formatted AnnData objects.\n",
    "dds.uns['trend_coeffs'] = np.array(dds.uns['trend_coeffs'])\n",
    "dds_gene.uns['trend_coeffs'] = np.array(dds_gene.uns['trend_coeffs'])\n",
    "\n",
    "dds.varm['replaced'] = np.array(dds.varm['replaced'])\n",
    "dds_gene.varm['replaced'] = np.array(dds_gene.varm['replaced'])\n",
    "\n",
    "# DeseqDataSet doesn't have native support for writing h5, save as AnnData objects and restore from\n",
    "# AnnData objects.\n",
    "dds.write(DDS_TRANSCRIPT_FH)\n",
    "dds_gene.write(DDS_GENE_FH)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pydeseq2",
   "language": "python",
   "name": "pydeseq2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.17"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
